graph/src/data/graphql/query_hash.rs

```diff
@@ -0,0 +1,455 @@
+use crate::prelude::q;
+use graphql_parser::query::{
+    Definition, Directive, Document, Field, FragmentDefinition, Mutation, Number,
+    OperationDefinition, Query, Selection, SelectionSet, Subscription, Text, Value,
+    VariableDefinition,
+};
+use graphql_tools::ast::{OperationTransformer, Transformed, TransformedValue};
+use std::cmp::Ordering;
+use std::collections::{hash_map::DefaultHasher, BTreeMap};
+use std::hash::{Hash, Hasher};
+use std::marker::PhantomData;
+
+pub fn query_hash(query: &q::Document) -> u64 {
```

Comment:
Unfortunately, `query_hash` needs to be at least a Level 2 hash: [http://nohatcoder.dk/2019-05-19-1.html](http://nohatcoder.dk/2019-05-19-1.html), keyed with a cryptographic-strength random number, and support a state space of at least `u128` for security.

The reason for this is that an insecure hash would allow a competent attacker to force an indexer to attest to an incorrect result by finding a hash collision.